### PR TITLE
copr/dag: add builtin conditions

### PR DESCRIPTION
**Pull request by choleraehyq**
@AndreMouche @andelf @BusyJay PTAL.

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_control.rs [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+use coprocessor::codec::Datum;
+use coprocessor::codec::mysql::{Decimal, Duration, Time};
+
+impl FnCall {
+    pub fn if_null_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let arg0 = try!(self.children[0].eval_int(ctx, row));
+        if !arg0.is_none() {
+            return Ok(arg0);
+        }
+        let arg1 = try!(self.children[1].eval_int(ctx, row));
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/builtin_control.rs]: I think `self.children[1].eval_int(ctx, row)` should be ok.

File: src/coprocessor/dag/expr/builtin_control.rs [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+        if !arg0.is_none() {
+            return Ok(arg0);
+        }
+        let arg1 = try!(self.children[1].eval_duration(ctx, row));
+        Ok(arg1)
+    }
+
+    pub fn if_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let arg0 = try!(self.children[0].eval_int(ctx, row));
+        let arg1 = try!(self.children[1].eval_int(ctx, row));
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/builtin_control.rs]: L92 and L93 should be moved to L96 and L95 respectively.

File: src/coprocessor/dag/expr/builtin_control.rs [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+        let arg2 = try!(self.children[2].eval_int(ctx, row));
+        match arg0 {
+            None | Some(0) => Ok(arg2),
+            _ => Ok(arg1),
+        }
+    }
+
+    pub fn if_real(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<f64>> {
+        let arg0 = try!(self.children[0].eval_int(ctx, row));
+        let arg1 = try!(self.children[1].eval_real(ctx, row));
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/builtin_control.rs]: Ditto.

File: src/coprocessor/dag/expr/builtin_control.rs [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+        }
+    }
+
+    pub fn if_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let arg0 = try!(self.children[0].eval_int(ctx, row));
+        let arg1 = try!(self.children[1].eval_decimal(ctx, row));
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/builtin_control.rs]: Ditto.

**Comment by choleraehyq** [comment]
@AndreMouche @hicqu @andelf PTAL

File: src/coprocessor/dag/expr/builtin_control.rs [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+    }
+
+    pub fn if_string<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Vec<u8>>>> {
+        let arg0 = try!(self.children[0].eval_int(ctx, row));
+        match arg0 {
+            None | Some(0) => Ok(try!(self.children[2].eval_string(ctx, row))),
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/builtin_control.rs]: `Ok(try!(` is redundant.

File: src/coprocessor/dag/expr/builtin_control.rs [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+    use coprocessor::codec::mysql::Duration;
+    use coprocessor::dag::expr::{Expression, StatementContext};
+    use coprocessor::dag::expr::test::{fncall_expr, str2dec};
+    use coprocessor::select::xeval::evaluator::test::datum_expr;
+
+    #[test]
+    fn test_if_null() {
+        let tests = vec![
+            (
+                ScalarFuncSig::IfNullInt,
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/builtin_control.rs]: Should check `Datum::I64(0)`.

**Comment by choleraehyq** [comment]
@AndreMouche PTAL

**Pull request merged**